platform-manageability-agent/internal/comms/comms.go

```diff
@@ -0,0 +1,183 @@
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/timeout"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/open-edge-platform/edge-node-agents/platform-manageability-agent/internal/config"
+	log "github.com/open-edge-platform/edge-node-agents/platform-manageability-agent/internal/logger"
+	pb "github.com/open-edge-platform/infra-external/dm-manager/pkg/api/dm-manager"
+)
+
+const retryInterval = 10 * time.Second
+const tickerInterval = 500 * time.Millisecond
+const connTimeout = 5 * time.Second
+
+type Client struct {
+	DMMgrServiceAddr string
+	Dialer           grpc.DialOption
+	Transport        grpc.DialOption
+	GrpcConn         *grpc.ClientConn
+	DMMgrClient      pb.DeviceManagementClient
+	RetryInterval    time.Duration
+}
+
+func WithNetworkDialer(serviceAddr string) func(*Client) {
+	return func(s *Client) {
+		s.Dialer = grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
+			return net.Dial("tcp", serviceAddr)
+		})
+	}
+}
+
+func NewClient(serviceURL string, tlsConfig *tls.Config) *Client {
+	cli := &Client{}
+	cli.DMMgrServiceAddr = serviceURL
+	cli.RetryInterval = retryInterval
+	cli.Transport = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+
+	WithNetworkDialer(cli.DMMgrServiceAddr)(cli)
+	return cli
+}
+
+func (cli *Client) Connect(ctx context.Context) (err error) {
+	cli.GrpcConn, err = grpc.DialContext(ctx, cli.DMMgrServiceAddr, cli.Transport, cli.Dialer, //nolint:staticcheck
+		grpc.WithUnaryInterceptor(timeout.UnaryClientInterceptor(connTimeout)),
+		grpc.WithStatsHandler(otelgrpc.NewClientHandler()))
+	if err != nil {
+		return fmt.Errorf("connection to %v failed: %v", cli.DMMgrServiceAddr, err)
+	}
+	cli.DMMgrClient = pb.NewDeviceManagementClient(cli.GrpcConn)
+	return nil
+}
+
+func ConnectToDMManager(ctx context.Context, serviceAddr string, tlsConfig *tls.Config) *Client {
+	dmMgr := NewClient(serviceAddr, tlsConfig)
+
+	cyclicalTicker := time.NewTicker(tickerInterval)
+	defer cyclicalTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Logger.Info("Connecting to DM Manager has been cancelled")
+			return nil
+		case <-cyclicalTicker.C:
+			err := dmMgr.Connect(ctx)
+			if err != nil {
+				log.Logger.Warnf("Can't connect to DM Manager, retrying: %v", err)
+				time.Sleep(dmMgr.RetryInterval)
+				continue
+			}
+			log.Logger.Info("Successfully connected to DM Manager")
+			return dmMgr
+		}
+	}
+}
+
+// Open Questions:
+// 1. I'm setting UUID in the hostID field, is this the correct approach or should we use a different identifier?
+// 2. If Version is empty, should we set it to a default value or leave it empty?
+// 3. If Version is not found in the output, should we simply set status to Disabled?
+// 4. Should we handle the case where rpc amtinfo command fails and send the error message to dm-manager?
```

Comment:
1. Yes host ID should be set to the UUID, as this is the unique identifier used to identify hosts in inventory
2. If version is empty, we should return it as empty rather than as a default value. The agent should report the configuration as it is on the host
3. If getting the version is what tells the agent whether AMT is supported, then yes it should return status Disabled when version is not found
4. If rpc command fails, then the error message should be recorded in the agent logs and a Disabled status sent to the dm-manager. We could also include a retry of the rpc command after a period of time to see if it then works later